JS機能実装（GIZFE-692）

Joex13 opened the pull request:
## チケットのリンク
https://gizumo.backlog.com/view/GIZFE-692

## やったこと
- JS系機能実装
  - ハンバーガーメニューの実装（SP）
  - スライダーの実装（Swiper使用）

## やらないこと
- smooth-behaivor: smooth;の指定（_base.scss）

## 特にレビューをお願いしたい箇所
なし

Joex13 committed:
feat: l-header, sp版のコーディングを作成

Joex13 committed:
fix: l-main, sp版でのl-headerとの余白調整

Joex13 committed:
feat: l-inner, sp版の実装※※※

※l-innerの仕様変更あり
width: 1180px→max-width: 1180px（PC）

Joex13 committed:
feat: p-key-visual, sp版仮実装（未完成）

Joex13 committed:
fix: 全体調整1

Joex13 committed:
feat: c-button, sp版の設定

Joex13 committed:
fix: l-inner, padding 0 15px設定

Joex13 committed:
fix: p-new-offers, 微修正

Joex13 committed:
fix: u-mt-xxx, sp版設定を記述

Joex13 committed:
fix: p-key-visual 未完

Joex13 committed:
fix: padding 0 15px 削除

Joex13 committed:
fix: p-useful, 実装修正

Joex13 committed:
fix: l-inner, padding 0 15px 復活

Joex13 committed:
fix: p-about-us, sp版完成

Joex13 committed:
fix: mixin heading-lv3にsp版を実装

Joex13 committed:
fix: p-flow, 三角形にsp版記述に変更

Joex13 committed:
fix: p-search, 微修正

Joex13 committed:
fix: 全体の余白調整

Joex13 committed:
fix: p-new-offers, sp版,listを中央寄せにfix

Joex13 committed:
fix: p-flow, sp版で常にpadding 0 60pxをかける

Joex13 committed:
fix: p-search, cssのミスを修正

sp版cssがpc版にも掛かってしまっていた

Joex13 committed:
fix: p-key-visual, 実装を変更

Joex13 committed:
fix: p-search, 修正

Joex13 committed:
fix: p-key-visual, 修正

Joex13 committed:
fix: l-header, pc版、aタグのwidthを設定

Joex13 committed:
fix: l-header, 左右のpaddingを変更（PC）

Joex13 committed:
fix: p-search, sp版を実装

Joex13 committed:
fix: p-flow, paddingの取り方を変更（SP/外見に変化なし）

Joex13 committed:
fix: 全体調整

Joex13 committed:
fix: tabの全体調整

Joex13 committed:
fix: 全体調整（tab版）

Joex13 committed:
fix: 全体修正（tab）

Joex13 committed:
fix: p-search, 暫定（tab）

Joex13 committed:
fix: key-visual

Joex13 committed:
fix: tab修正

Joex13 committed:
fix: htmlエラー修正

Joex13 committed:
fix: p-features, mt55px!importantを設定

Joex13 committed:
fix: 課題→headerが崩れる

Joex13 committed:
fix: tab, test

Joex13 committed:
fix: rem化

Joex13 committed:
fix: btn修正

Joex13 committed:
fix: btn修正2

Joex13 committed:
fix: padding-fix

Joex13 committed:
fix: footer-fix

Joex13 committed:
fix: 全体修正

Joex13 committed:
fix about

Joex13 committed:
fix: l-header, 余白fix

Joex13 committed:
fix: max-width:auto→initial;

Joex13 committed:
fix: heightを削除(l-header)

Joex13 committed:
fix: rem化

Joex13 committed:
fix: rem化

Joex13 committed:
fix: header-nav系 padding

Joex13 committed:
fix: search レイアウト崩れの修正

Joex13 committed:
fix: button レイアウトの修正

Joex13 committed:
fix: key-visual, heightをレスポンシブ化

Joex13 committed:
fix: 全体余白修正,search-bottomの復活

Joex13 committed:
fix: font-feature-settings: 'palt';を設定（sp実機表示崩れ対策）

Joex13 committed:
fix: 全体修正

Joex13 committed:
fix: search, 全体修正

Joex13 committed:
fix: 全体修正

Joex13 committed:
fix: id被りを修正（html）

Joex13 committed:
fix: font-sizeの微修正

Joex13 committed:
style: 可読性向上&font-sizeのミス修正

Joex13 committed:
style: 可読性向上fix

Joex13 committed:
fix: ▲の位置を修正

Joex13 committed:
fix: 全体修正

Joex13 committed:
fix: layoutのレスポンシブ化

Joex13 committed:
fix: 縦横比崩れ（key-visual）

Joex13 committed:
fix: searchのrem化

Joex13 committed:
fix: rem化（new-offers）

Joex13 committed:
fix: rem化（new-offers）

Joex13 committed:
fix: rem化（features）

Joex13 committed:
fix: rem化（flow/不具合あり）

Joex13 committed:
fix: rem化(useful)

Joex13 committed:
fix: rem化(flow)

Joex13 committed:
fix: rem化(new-offers)

Joex13 committed:
fix: buttonのレイアウト崩れの修正

Joex13 committed:
fix: rem化（utility）

Joex13 committed:
fix: rem化(utility)

Joex13 committed:
fix: 不要なクラスの削除

Joex13 committed:
fix: 疑似要素の修正（search）

Joex13 committed:
fix: buttonのcssを修正

Joex13 committed:
fix: 指定漏れ修正

Joex13 committed:
fix: rem化（hamburger）

Joex13 committed:
fix: key-visualの"求人掲載数"をリンク化

Joex13 committed:
fix: brをコメントアウト

Joex13 committed:
fix レビュー修正（山口メンター指摘分）

Joex13 committed:
fix: レビュー修正漏れ（山口メンター指摘分）

Joex13 committed:
feat: ハンバーガーメニューの表示トグルまで実装

Joex13 committed:
fix: ハンバーガーメニューをXにする処理の実装

Joex13 committed:
fix: スクロールロックバグを仮修正（画面サイズ増加時）

Joex13 committed:
chore: Swiperの導入

Joex13 committed:
feat: Swiperの初期導入完了（動きます）

Joex13 committed:
test: Swiperのnavigationアイコンを変えようと試行錯誤中

Joex13 committed:
test: デフォルトデザインのnavigation.ver

Joex13 committed:
fix: 横スクロールが出来ていたバグを修正

Joex13 committed:
fix: package,package-lockが変更される問題の暫定処置

Joex13 committed:
fix: コンフリクトを解消？

Joex13 committed:
コンフリクト解消

Joex13 committed:
fix: コンフリクト解消操作のミス修正

Joex13 committed:
feat: スクロールロック仮

Joex13 committed:
fix: vanillaJS化

Joex13 committed:
style: コメントの修正

Joex13 committed:
fix: スムーススクロールを実装

Joex13 committed:
fix: slick-carouselの削除

keisuke3 commented on src/css/foundation/_base.scss:
```diff
@@ -11,6 +11,7 @@
 html {
   font-size: 10px;
   font-feature-settings: 'palt'; // sp用。footer-navの改行対策。
+  scroll-behavior: smooth;
```
[must]
スクロールした時にタイトルがヘッダーに隠れているので、ヘッダーが隠れないように修正をお願いします！

Uploading 画面収録 2023-03-22 11.23.09.mov…

jj13jp replied:
修正致しました！

keisuke3 commented on src/css/foundation/_base.scss:
```diff
@@ -11,6 +11,7 @@
 html {
   font-size: 10px;
   font-feature-settings: 'palt'; // sp用。footer-navの改行対策。
+  scroll-behavior: smooth;
```
[must]
ハンバーガーメニュー内のリンクをクリックした時、リンク箇所にスクロールすると思うのですが、ハンバーガーメニューが開きっぱなしになってます！
リンクをクリックした時はハンバーガーメニューが閉じるようにお願いします！

jj13jp replied:
修正致しました！

keisuke3 commented on src/index.html:
```diff
@@ -127,13 +127,15 @@
           </ul>
         </nav>
-        <div class="c-hamburger">
+        <div class="c-hamburger is-sp">
```
[must]
ハンバーガーメニューはdivタグではなくbuttonタグで作るようにお願いします！
divタグで作るとキーボードでフォーカスが当たらず操作できないなど、ボタンとしての機能を満たさないものになってしまいます。（buttonタグを使う理由は他にもいくつかあります）

jj13jp replied:
修正致しました！

keisuke3 commented on src/css/javascript/swiper.scss:
```diff
@@ -0,0 +1,138 @@
+
+  // &-button-prev::after {
+  //   content: '';
+  // }
+
+  // &-button-prev > span {
+  //   content: '';
+  //   position: absolute;
+  //   width: 20px;
+  //   height: 0.7px;
+  //   background: vm.$black;
+  //   @include vm.tab {
+  //     width: 0.2rem;
+  //     height: 0.007rem;
+  //   }
+  //   @include vm.sp {
+  //     width: 0.2rem;
+  //     height: 0.007rem;
+  //   }
+  // }
+
+  // &-button-prev > span:first-child {
+  //   transform: translateY(-7px) rotate(-45deg);
+  //   @include vm.tab {
+  //     transform: translateY(-0.07rem) rotate(-45deg);
+  //   }
+  //   @include vm.sp {
+  //     transform: translateY(-0.07rem) rotate(-45deg);
+  //   }
+  // }
+
+  // &-button-prev > span:last-child {
+  //   transform: translateY(7px) rotate(45deg);
+  //   @include vm.tab {
+  //     transform: translateY(0.07rem) rotate(45deg);
+  //   }
+  //   @include vm.sp {
+  //     transform: translateY(0.07rem) rotate(45deg);
+  //   }
+  // }
```
[must]
不要な記述であればコメントアウトではなく削除をお願いします！
他にもコードのコメントアウトは基本的に削除するようにお願いします！

jj13jp replied:
修正致しました！

keisuke3 commented on src/index.html:
```diff
@@ -127,13 +127,15 @@
-            <li>あああ</li>
-            <li>あああ</li>
+          <ul class="l-header__nav__list--sp is-sp">
+            <li><a href="#top">トップ</a></li>
+            <li><a href="#new-offers">新着求人</a></li>
+            <li><a href="#features">4つの特徴</a></li>
+            <li><a href="#flow">転職までの流れ</a></li>
+            <li><a href="#useful">転職お役立ちコンテンツ</a></li>
```
[imo]
ハンバーガーメニュー内のリンクのクリックエリアを広くしてあげた方が良いかなと思います！
スマホ画面でタップすることが多いと思うので、現状の広さだとタップしにくいかなと思います！

<img width="364" alt="スクリーンショット 2023-03-22 11 29 10" src="https://user-images.githubusercontent.com/45491269/226786892-b503c981-6488-4f67-8195-b680446a4b83.png">

jj13jp replied:
修正致しました！

keisuke3 reviewed (commented):
コメントしましたので確認お願いします！

Joex13 committed:
style: 不要なコメントを削除

レビュー修正分

Joex13 committed:
fix: ハンバーガーメニュー内のリンクのタップエリアを広げた

レビュー修正分

Joex13 committed:
fix: ハンバーガーメニューをdiv→buttonに変更

レビュー修正分

Joex13 committed:
fix: ハンバーガーメニュー内のリンク時、ハンバーガーメニューを閉じるように変更&リファクタリング

レビュー修正分

Joex13 committed:
fix: jQueryのimport記述を削除

Joex13 committed:
fix: ページ内ジャンプ時の位置問題を修正

レビュー修正分

Joex13 committed:
fix: スライダーアイテム間の余白を修正（20→27）

sp最大サイズ付近で、左右のアイテムが見えてしまっていた

Joex13 committed:
fix: ハンバーガーメニューが細くなっていた問題を修正

ブラウザデフォルトのpadding指定を外した

Joex13 committed:
fix: ページ内ジャンプのアンカーをa→spanに変更

SP実機での挙動がおかしかったため

jj13jp commented:
## 修正した点は以下の通り
- [ ] ハンバーガーメニュー内のリンク押下時、スクロール先のheading要素がheaderの下に隠れる問題を修正する
- [ ] ハンバーガーメニュー内のリンク押下時、ハンバーガーメニューが閉じるようにする
- [ ] ハンバーガーメニューに使用するタグをdiv→buttonにする
- [ ] コメントアウト済みのコードを削除する
- [ ] ハンバーガーメニュー内のリンクのクリックエリアを拡張する

Joex13 committed:
refactor: isScrollableのboolean反転処理を最適化

Joex13 committed:
refactor: コメントのミスを修正

keisuke3 commented on src/index.html:
```diff
@@ -308,6 +395,7 @@ <h2 class="p-about-us__heading">
           </h2>
         </article>
         <article class="p-features l-inner u-mt-large">
+          <span href="" id="features" class="u-anchor"></span>
```
[imo]
アンカーリンク用のspanタグは必要ないかなと思うのですが、いかがでしょうか？
一つ上のarticleにidを設定すればスクロールは実現できそうです！

jj13jp replied:
こちら修正致しました！

### spanタグを用いた上記のような実装は避けるべきでしょうか？
その場合、なぜ避けるべきなのでしょうか？

Joex13 replied:
こちら修正致しました！

keisuke3 replied:
@Joex13 
シンプルにspanを追加して回りくどい実装をする必要がないからですね！
後から別の実装者がコードを見た時に、アンカーリンク用のspanタグがあると「何のためにこうしてるんだ？」という疑問がおそらくほとんどの実装者の頭によぎると思います。

このような何のためにそうなってるか分からないコードは、デグレによるバグが怖いので改修も億劫になりやすいです。
なのでなるべく避けた方が良いかなと思いました！

keisuke3 commented on src/js/app.js:
```diff
@@ -1,5 +1,66 @@
+    document
+      .getElementsByClassName('l-header__nav__list--sp')[0]
+      .classList.toggle('js-active');
+    if (isScrollable) {
+      document.addEventListener('wheel', banScroll, { passive: false });
+      document.addEventListener('touchmove', banScroll, { passive: false });
+    } else {
+      document.removeEventListener('wheel', banScroll);
+      document.removeEventListener('touchmove', banScroll);
+    }
```
[ask]
ここの処理はどういったことを想定して書いてますか？？

jj13jp replied:
変数isScrollableの値（true or false）に応じて、
スクロール禁止の設定と解除をそれぞれで行っています。

"wheel"はpcユーザーを、"touchmove"はspユーザーを想定した実装です。

keisuke3 reviewed (commented):
コメントしましたので確認お願いします！

Joex13 committed:
fix: スクロール禁止時にスクロールバーも非表示に

Joex13 committed:
fix: ページ内リンクの実装を変更

レビュー修正分
spanタグを削除、親要素のarticle,divに指定するように
:target {scroll-margin-top: 0.7rem;}も設定

Joex13 committed:
refactor: 未使用のクラスを削除

keisuke3 reviewed (commented):
これは修正事項ではないのですが、commitの粒度が小さすぎるのでもう少し大きな粒度でcommitしましょう！
レビュアーが細かすぎてcommit単位でコードを見ようという気にならないかなと思います。
（commit単位で見たとしても細かすぎて意味のあるまとまりになっていない）
